### PR TITLE
Revise OneHot operation reference implementation

### PR DESCRIPTION
**Pull request by IvanNovoselov**
This is a revision of OneHot reference implementation and corresponding nGraph operator class. Please, note that:
-  OneHot documentation was updated in https://github.com/openvinotoolkit/openvino/pull/4243
- OneHot single-layer tests were implemented in https://github.com/openvinotoolkit/openvino/pull/4059
- Arm_plugin calls reference impl of OneHot directly. Therefore a separate [PR #52](https://github.com/openvinotoolkit/openvino_contrib/pull/52) was submitted to openvino_contrib in order to update the plugin.

Ticket: 37522

### CODE REVIEW

**Comment by iefode** [comment]
@lazarevevgeny @ilyachur @jdanieck Could you please take a look?

File: ngraph/core/reference/include/ngraph/runtime/reference/one_hot.hpp
Comment (IvanNovoselov) [reply ngraph/core/reference/include/ngraph/runtime/reference/one_hot.hpp:29]: 1) A proper output shape is set in [validate_and_infer_types()](https://github.com/openvinotoolkit/openvino/blob/0d2cfdb866e40398ef037abadae570fc5d66b8f8/ngraph/core/src/op/one_hot.cpp#L107-L110). So if we assume that `validate_and_infer_types()` was called before `op::v1::OneHot::evaluate(...)`, it seems that we don't have to pass the whole output shape (which is exactly the input shape + depth inserted at the axis position). Or you believe it is better to re-check the shapes compliance inside the `reference::one_hot<...>(...)`? Like `( shape_size(out_shape) == shape_size(indices_shape) * depth )` for example?
2) Please, note that `( shape_size(out_shape) == shape_size(indices_shape) * depth )` must hold at the time when `reference::one_hot<...>(...)` is called. Therefore `out + (num_ind * depth - 1) * out_elem_size)`  could point outside the allocated output memory block, only if an incorrect `out_elem_size` was provided. More precisely, if `out_elem_size > output_values[0]->get_element_type().size()`. However, this should not be the case, when `reference::one_hot<...>(...)` is called from the [op::v1::OneHot::evaluate(...)](https://github.com/openvinotoolkit/openvino/blob/0d2cfdb866e40398ef037abadae570fc5d66b8f8/ngraph/core/src/op/one_hot.cpp#L142-L149). It is not clear, whether we can easily check that the `out_elem_size` is correct without providing output element type explicitly as a template argument.

Comment (ilyachur) [reply ngraph/core/reference/include/ngraph/runtime/reference/one_hot.hpp:29]: I think we that we need to have such checks in the evaluate method or here. It should help to avoid any SEGFAULTs for incorrect usages.

Comment (IvanNovoselov) [reply ngraph/core/reference/include/ngraph/runtime/reference/one_hot.hpp]: @ilyachur, agreed, thank you! 
Added a check for I/O shapes consistency:
https://github.com/openvinotoolkit/openvino/blob/3a5eded6af710402ef8de2482af0f03b65313bd5/ngraph/core/reference/include/ngraph/runtime/reference/one_hot.hpp#L34-L39

The check is performed inside `reference::one_hot<...>(...)`, since potentially it could be called directly through evaluates_map (thus bypassing the `evaluate(...)` method). Please, correct me, if I getting smth wrong here.

File: ngraph/test/constant_folding.cpp
Comment (jdanieck) [inline ngraph/test/constant_folding.cpp:2422]: This test name is `constant_v1_one_hot_negative_axes` which suggest me that test was intended for negative axes testing, which suggest they should be supported.  I see that you change spec and said explicitly that negative indexes are not supported,  you consulted this change with someone right ? :) If so, then I think we can remove this test at all, or at least rename it - now it is confusing.

File: inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/one_hot.cpp
Comment (IvanNovoselov) [reply inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/one_hot.cpp:12]: Ok. Done!

File: ngraph/test/constant_folding.cpp
Comment (IvanNovoselov) [reply ngraph/test/constant_folding.cpp:2422]: @jdanieck, you are right, the behavior for negative indices is undefined according to the new version of specs (please, see the [discussion](https://github.com/openvinotoolkit/openvino/pull/4243#issuecomment-778197662) for details). However, negative axis values are still allowed and supported (please, see the [specs](https://github.com/openvinotoolkit/openvino/blame/c001e54403c3ead20f6fc4cf96b290af00895002/docs/ops/sequence/OneHot_1.md#L25-L29) and [implementation](https://github.com/openvinotoolkit/openvino/blob/3a5eded6af710402ef8de2482af0f03b65313bd5/ngraph/core/src/op/one_hot.cpp#L84-L85)).

File: ngraph/test/runtime/interpreter/evaluates_map.cpp
Comment (IvanNovoselov) [reply ngraph/test/runtime/interpreter/evaluates_map.cpp]: Ok, I will remove one_hot from evaluates_map.cpp then.

File: ngraph/core/reference/include/ngraph/runtime/reference/one_hot.hpp
Comment (IvanNovoselov) [reply ngraph/core/reference/include/ngraph/runtime/reference/one_hot.hpp:29]: Moving the checks to the `evaluate(...)` method, since `reference::one_hot<...>(...)` could not be called from evaluates_map according to [this comment](https://github.com/openvinotoolkit/openvino/pull/4330#discussion_r578271336).

File: ngraph/test/runtime/interpreter/evaluates_map.cpp
Comment (IvanNovoselov) [reply ngraph/test/runtime/interpreter/evaluates_map.cpp]: @jdanieck, several CI jobs fail for arm after removing one_hot from evaluates_map (more details [here](https://openvino-ci.intel.com/job/private-ci/job/github_trigger/job/openvino/job/PR-4330/6/)) :
`build arm-plugin ERROR: /home/jenkins/agent/workspace/private-ci/ie/build-linux-android_arm/b/repos/openvino_contrib/modules/arm_plugin/src/arm_converter/arm_converter_one_hot.cpp:18:5: error: no matching function for call to 'one_hot'`
So I guess we do need `evaluate` in evaluates_map as some kind of fallback.

**Comment by IvanNovoselov** [comment]
@ilyachur @jdanieck please take a second look and let me know if you have any further concerns.

**Comment by IvanNovoselov** [comment]
Please note that several CI jobs (android_arm, debian_9_arm, yocto_kmb and ie_tests_cpu_arm_debian_9_arm) failed because arm_plugin calls reference OneHot implementation directly (not through evaluate() of the op class), and the signature of `reference::one_hot<>(...)` has changed in this PR. Therefore a separate [PR #52](https://github.com/openvinotoolkit/openvino_contrib/pull/52) was submitted to openvino_contrib in order to fix arm_plugin.

**Comment by IvanNovoselov** [comment]
@ilyachur could you please approve this PR if there are no further suggestions?

**Review by ilyachur — APPROVED** [review]
LGTM, please resolve merge conflicts.

**Review by jdanieck — APPROVED** [review]

**Comment by ilyachur** [comment]
@IvanNovoselov  Please set the milestone.

**Comment by IvanNovoselov** [comment]
> @IvanNovoselov Please set the milestone.

@ilyachur, done. The milestone 2021.4 is set.

**Comment by IvanNovoselov** [comment]
@ilyachur @lazarevevgeny please merge, if everything is Ok.  Also note [this comment](https://github.com/openvinotoolkit/openvino/pull/4330#issuecomment-786019870) on why some tests fail.

**Comment by ilyachur** [comment]
@lazarevevgeny Please have a look.
@IvanNovoselov We cannot merge it, while CI is red.

**Comment by azhogov** [comment]
CI passed: https://gitlab-icv.inn.intel.com/inference-engine/product-configs/merge_requests/1430

**Pull request merged**
